File: app/src/main/java/tech/salroid/filmy/Activity/MainActivity.java
package tech.salroid.filmy.activity;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.speech.RecognizerIntent;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.crashlytics.android.Crashlytics;
import com.miguelcatalan.materialsearchview.MaterialSearchView;

import java.util.ArrayList;

import io.fabric.sdk.android.Fabric;
import tech.salroid.filmy.R;
import tech.salroid.filmy.custom_adapter.MyPagerAdapter;
import tech.salroid.filmy.fragments.InTheaters;
import tech.salroid.filmy.fragments.SearchFragment;
import tech.salroid.filmy.fragments.Trending;
import tech.salroid.filmy.fragments.UpComing;
import tech.salroid.filmy.service.FilmyJobScheduler;
import tech.salroid.filmy.utils.Network;
import tr.xip.errorview.ErrorView;


public class MainActivity extends AppCompatActivity {


    public boolean fetchingFromNetwork;
    private MaterialSearchView materialSearchView;
    private SearchFragment searchFragment;
    private ViewPager viewPager;
    private TabLayout tabLayout;
    private ErrorView mErrorView;
    private Trending trendingFragment;
    private FrameLayout toolbarScroller;
    private boolean cantProceed;
    private BroadcastReceiver mMessageReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            // Extract data included in the Intent
            int statusCode = intent.getIntExtra("message",00);

            Toast.makeText(context,"Failed to get latest movies.",Toast.LENGTH_SHORT).show();

            cantProceed(statusCode);

        }
    };

    @Override

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Fabric.with(this, new Crashlytics());
        setContentView(R.layout.activity_main);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        if (getSupportActionBar()!=null)
        getSupportActionBar().setTitle(" ");




        TextView logo = (TextView) findViewById(R.id.logo);

        toolbarScroller = (FrameLayout) findViewById(R.id.toolbarScroller);

        Typeface typeface = Typeface.createFromAsset(getAssets(), "fonts/canaro_extra_bold.otf");
        logo.setTypeface(typeface);

        mErrorView = (ErrorView) findViewById(R.id.error_view);

        mErrorView.setConfig(ErrorView.Config.create()
                .title(getString(R.string.error_title_damn))
                .titleColor(ContextCompat.getColor(this,R.color.dark))
                .subtitle("Unable to fetch movies.\nCheck internet connection then try again.")
                .retryText(getString(R.string.error_view_retry))
                .build());


        mErrorView.setOnRetryListener(new ErrorView.RetryListener() {
            @Override
            public void onRetry() {

                if (Network.isNetworkConnected(MainActivity.this)) {

                    fetchingFromNetwork = true;
                    setScheduler();

                }

                canProceed();

            }
        });



        mErrorView.setVisibility(View.GONE);

        // Get the ViewPager and set it's PagerAdapter so that it can display items
        viewPager = (ViewPager) findViewById(R.id.viewpager);
        setupViewPager(viewPager);

        // Give the TabLayout the ViewPager
        tabLayout = (TabLayout) findViewById(R.id.tab_layout);
        tabLayout.setupWithViewPager(viewPager);

        materialSearchView = (MaterialSearchView) findViewById(R.id.search_view);
        materialSearchView.setVoiceSearch(true);
        materialSearchView.setOnQueryTextListener(new MaterialSearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                //Do some magic

                getSearchedResult(query);
                searchFragment.showProgress();

                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {

                //getSearchedResult(newText);
                return true;

            }
        });

        materialSearchView.setOnSearchViewListener(new MaterialSearchView.SearchViewListener() {
            @Override
            public void onSearchViewShown() {
                //Do some magic

                tabLayout.setVisibility(View.GONE);

                disableToolbarScrolling();

                searchFragment = new SearchFragment();
                getSupportFragmentManager().
                        beginTransaction().
                        replace(R.id.search_container, searchFragment)
                        .commit();

            }

            @Override
            public void onSearchViewClosed() {
                //Do some magic


                getSupportFragmentManager()
                        .beginTransaction()
                        .remove(searchFragment)
                        .commit();

                if (!cantProceed)
                  tabLayout.setVisibility(View.VISIBLE);

                enableToolbarScrolling();

            }
        });



        if (Network.isNetworkConnected(this)) {

            fetchingFromNetwork = true ;
            setScheduler();
        }

    }

    private void setScheduler() {

        FilmyJobScheduler filmyJobScheduler = new FilmyJobScheduler(this);
        filmyJobScheduler.createJob();

    }

    public void cantProceed(final int status) {

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {

                cantProceed = true;

                tabLayout.setVisibility(View.GONE);
                viewPager.setVisibility(View.GONE);
                mErrorView.setError(status);
                mErrorView.setVisibility(View.VISIBLE);

                //disable toolbar scrolling
                disableToolbarScrolling();

            }
        }, 1000);

    }

    public void canProceed(){

        cantProceed = false;

        tabLayout.setVisibility(View.VISIBLE);
        viewPager.setVisibility(View.VISIBLE);
        mErrorView.setVisibility(View.GONE);

        if(trendingFragment!=null){

            trendingFragment.retryLoading();

        }

        enableToolbarScrolling();

    }

    private void disableToolbarScrolling() {
        AppBarLayout.LayoutParams params = (AppBarLayout.LayoutParams) toolbarScroller.getLayoutParams();
        params.setScrollFlags(0);
    }

    private void enableToolbarScrolling() {

        AppBarLayout.LayoutParams params =
                (AppBarLayout.LayoutParams) toolbarScroller.getLayoutParams();
        params.setScrollFlags(AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                | AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS);

    }

    private void setupViewPager(ViewPager viewPager) {


        trendingFragment = new Trending();
        InTheaters inTheatersFragment = new InTheaters();
        UpComing upComingFragment = new UpComing();

        MyPagerAdapter adapter = new MyPagerAdapter(getSupportFragmentManager());
        adapter.addFragment(trendingFragment, "TRENDING");
        adapter.addFragment(inTheatersFragment, "IN THEATERS");
        adapter.addFragment(upComingFragment, "UPCOMING");
        viewPager.setAdapter(adapter);

    }

    private void getSearchedResult(String query) {

        if (searchFragment != null) {
            searchFragment.getSearchedResult(query);
        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {


        getMenuInflater().inflate(R.menu.main_menu, menu);
        MenuItem item = menu.findItem(R.id.action_search);
        materialSearchView.setMenuItem(item);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_search) {

            startActivity(new Intent(this, SearchFragment.class));
        }

        if (id == R.id.ic_setting) {

                throw new RuntimeException("This is a crash");

           // startActivity(new Intent(this, SettingsActivity.class));

        }

        if (id == R.id.ic_collection) {
            startActivity(new Intent(this, SavedMovies.class));
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == MaterialSearchView.REQUEST_VOICE && resultCode == RESULT_OK) {
            ArrayList<String> matches = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS);
            if (matches != null && matches.size() > 0) {
                String searchWrd = matches.get(0);
                if (!TextUtils.isEmpty(searchWrd)) {
                    materialSearchView.setQuery(searchWrd, false);
                }
            }

            return;
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onBackPressed() {
        if (materialSearchView.isSearchOpen()) {
            materialSearchView.closeSearch();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        LocalBroadcastManager.getInstance(this).registerReceiver(mMessageReceiver,
                new IntentFilter("fetch-failed"));

    }

    @Override
    protected void onPause() {
        // Unregister since the activity is not visible
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mMessageReceiver);
        super.onPause();
    }



}

File: app/src/main/java/tech/salroid/filmy/ui/activities/MainActivity.kt
package tech.salroid.filmy.ui.activities

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.speech.RecognizerIntent
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.preference.PreferenceManager
import androidx.viewpager.widget.ViewPager
import com.google.android.material.appbar.AppBarLayout
import com.miguelcatalan.materialsearchview.MaterialSearchView
import com.miguelcatalan.materialsearchview.MaterialSearchView.SearchViewListener
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import tech.salroid.filmy.R
import tech.salroid.filmy.data.network.NetworkUtil
import tech.salroid.filmy.databinding.ActivityMainBinding
import tech.salroid.filmy.ui.FilmyIntro
import tech.salroid.filmy.ui.activities.fragment.InTheaters
import tech.salroid.filmy.ui.activities.fragment.Trending
import tech.salroid.filmy.ui.activities.fragment.UpComing
import tech.salroid.filmy.ui.adapters.CollectionsPagerAdapter
import tech.salroid.filmy.ui.fragment.SearchFragment
import tech.salroid.filmy.utility.getQueryTextChangeStateFlow

class MainActivity : AppCompatActivity() {

    var fetchingFromNetwork = false
    private var trendingFragment: Trending? = null
    private lateinit var searchFragment: SearchFragment
    private var cantProceed = false
    private var nightMode = false
    private lateinit var binding: ActivityMainBinding

    private val mMessageReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val statusCode = intent.getIntExtra("message", 0)
            cantProceed(statusCode)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        val sp = PreferenceManager.getDefaultSharedPreferences(this)
        nightMode = sp.getBoolean("dark", false)
        if (nightMode) setTheme(R.style.AppTheme_Base_Dark) else setTheme(R.style.AppTheme_Base)
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)
        supportActionBar?.title = " "
        introLogic()
        if (nightMode) allThemeLogic() else lightThemeLogic()

       /* binding.mainErrorView.apply {
            title = getString(R.string.error_title_damn)
            titleColor = ContextCompat.getColor(context, R.color.dark)
            setSubtitle(getString(R.string.error_details))
            setRetryText(getString(R.string.error_retry))
        }*/

        /*binding.mainErrorView.setRetryListener {
            if (NetworkUtil.isNetworkConnected(this@MainActivity)) {
                fetchingFromNetwork = true
                // fetchMoviesFromNetwork()
            }
            canProceed()
        }*/

        //binding.mainErrorView.visibility = View.GONE

        setupSearch()
    }

    private fun setupSearch() {
        binding.searchView.setVoiceSearch(false)
        binding.searchView.setOnSearchViewListener(object : SearchViewListener {
            override fun onSearchViewShown() {
                binding.tabLayout.visibility = View.GONE
                disableToolbarScrolling()
                searchFragment = SearchFragment()
                supportFragmentManager.beginTransaction()
                    .replace(R.id.search_container, searchFragment)
                    .commit()
            }

            override fun onSearchViewClosed() {
                supportFragmentManager
                    .beginTransaction()
                    .remove(searchFragment)
                    .commit()
                enableToolbarScrolling()
            }
        })

        // Instant search using Flow
        lifecycleScope.launch(Dispatchers.IO) {
            binding.searchView.getQueryTextChangeStateFlow()
                .debounce(300)
                .filter { query ->
                    return@filter query.isNotEmpty()
                }
                .distinctUntilChanged()
                .flatMapLatest { query ->
                    searchFragment.showProgress()
                    flow {
                        emit(NetworkUtil.searchMovies(query))
                    }.catch {
                        emitAll(flowOf(null))
                    }
                }
                .flowOn(Dispatchers.Main)
                .collect { result ->
                    lifecycleScope.launch(Dispatchers.Main) {
                        result?.results?.let { searchFragment.showSearchResults(it) }
                    }
                }
        }
    }

    private fun allThemeLogic() {
        binding.tabLayout.setTabTextColors(Color.parseColor("#bdbdbd"), Color.parseColor("#e0e0e0"))
        binding.tabLayout.setBackgroundColor(
            ContextCompat.getColor(
                this,
                R.color.colorDarkThemePrimary
            )
        )
        binding.tabLayout.setSelectedTabIndicatorColor(Color.parseColor("#bdbdbd"))
        binding.logo.setTextColor(Color.parseColor("#E0E0E0"))
        binding.searchView.setBackgroundColor(resources.getColor(R.color.colorDarkThemePrimary))
        binding.searchView.setBackIcon(
            ContextCompat.getDrawable(
                this,
                R.drawable.ic_action_navigation_arrow_back_inverted
            )
        )
        binding.searchView.setCloseIcon(
            ContextCompat.getDrawable(
                this,
                R.drawable.ic_action_navigation_close_inverted
            )
        )
        binding.searchView.setTextColor(Color.parseColor("#ffffff"))
    }

    private fun lightThemeLogic() {
        binding.logo.setTextColor(ContextCompat.getColor(this, R.color.dark))
    }

    private fun introLogic() {
        val thread = Thread {
            val getPrefs = PreferenceManager.getDefaultSharedPreferences(baseContext)
            val isFirstStart = getPrefs.getBoolean("firstStart", true)
            if (isFirstStart) {
                val i = Intent(this@MainActivity, FilmyIntro::class.java)
                startActivity(i)
                val e = getPrefs.edit()
                e.putBoolean("firstStart", false)
                e.apply()
            }
        }
        thread.start()
    }

    fun cantProceed(status: Int) {
        Handler().postDelayed({
            if (trendingFragment != null && !trendingFragment!!.isShowingFromDatabase) {
                cantProceed = true
                binding.tabLayout.visibility = View.GONE
                //binding.viewpager.visibility = View.GONE
                //binding.mainErrorView.visibility = View.VISIBLE
                disableToolbarScrolling()
            }
        }, 1000)
    }

    private fun canProceed() {
        cantProceed = false
        binding.tabLayout.visibility = View.VISIBLE
        //binding.viewpager.visibility = View.VISIBLE
        //binding.mainErrorView.visibility = View.GONE
        //trendingFragment?.retryLoading()
        enableToolbarScrolling()
    }

    private fun disableToolbarScrolling() {
        val params = binding.toolbarScroller.layoutParams as AppBarLayout.LayoutParams
        params.scrollFlags = 0
    }

    private fun enableToolbarScrolling() {
        val params = binding.toolbarScroller.layoutParams as AppBarLayout.LayoutParams
        params.scrollFlags = (AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                or AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS)
    }

    private fun setupViewPager(viewPager: ViewPager) {
        trendingFragment = Trending()
        val inTheatersFragment = InTheaters()
        val upComingFragment = UpComing()

        val adapter = CollectionsPagerAdapter(supportFragmentManager)
        adapter.addFragment(trendingFragment!!, getString(R.string.trending))
        // adapter.addFragment(inTheatersFragment, getString(R.string.theatres))
        // adapter.addFragment(upComingFragment, getString(R.string.upcoming))
        viewPager.adapter = adapter
    }

    private fun getSearchedResult(query: String) {
        searchFragment.getSearchedResult(query)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        val itemSearch = menu.findItem(R.id.action_search)
        val itemAccount = menu.findItem(R.id.ic_collections)

        if (nightMode) {
            itemSearch.setIcon(R.drawable.ic_action_action_search)
            itemAccount.setIcon(R.drawable.ic_action_collections_bookmark2)
        }

        binding.searchView.setMenuItem(itemSearch)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_search -> startActivity(Intent(this, SearchFragment::class.java))
            R.id.ic_setting -> startActivity(Intent(this, SettingsActivity::class.java))
            R.id.ic_collections -> startActivity(Intent(this, CollectionsActivity::class.java))
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == MaterialSearchView.REQUEST_VOICE && resultCode == RESULT_OK) {
            val matches = data!!.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
            if (matches != null && matches.size > 0) {
                val searchWrd = matches[0]
                if (!TextUtils.isEmpty(searchWrd)) {
                    binding.searchView.setQuery(searchWrd, false)
                }
            }
            return
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onBackPressed() {
        if (binding.searchView.isSearchOpen) {
            binding.searchView.closeSearch()
        } else {
            super.onBackPressed()
        }
    }

    override fun onResume() {
        super.onResume()
        val sp = PreferenceManager.getDefaultSharedPreferences(this)
        val nightModeNew = sp.getBoolean("dark", false)
        if (nightMode != nightModeNew) recreate()
        LocalBroadcastManager.getInstance(this)
            .registerReceiver(mMessageReceiver, IntentFilter("fetch-failed"))
    }

    override fun onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mMessageReceiver)
        super.onPause()
    }
}